mllib/src/main/scala/org/apache/spark/ml/stat/ANOVATest.scala

```diff
@@ -0,0 +1,166 @@
+   * @param labelCol  Name of label column in dataset, of any numerical type
+   * @return DataFrame containing the test result for every feature against the label.
+   *         This DataFrame will contain a single Row with the following fields:
+   *          - `pValues: Vector`
+   *          - `degreesOfFreedom: Array[Long]`
+   *          - `fValues: Vector`
+   *         Each of these fields has one value per feature.
+   */
+  @Since("3.1.0")
+  def test(dataset: DataFrame, featuresCol: String, labelCol: String): DataFrame = {
+    val spark = dataset.sparkSession
+    val testResults = testClassification(dataset, featuresCol, labelCol)
+    val pValues: Vector = Vectors.dense(testResults.map(_.pValue))
+    val degreesOfFreedom: Array[Long] = testResults.map(_.degreesOfFreedom)
+    val fValues: Vector = Vectors.dense(testResults.map(_.statistic))
+    spark.createDataFrame(
+      Seq(new ANOVAResult(pValues, degreesOfFreedom, fValues)))
+  }
+
+  /**
+   * @param dataset  DataFrame of categorical labels and continuous features.
+   * @param featuresCol  Name of features column in dataset, of type `Vector` (`VectorUDT`)
+   * @param labelCol  Name of label column in dataset, of any numerical type
+   * @return Array containing the ANOVATestResult for every feature against the
+   *         label.
+   */
+  private[ml] def testClassification(
+      dataset: Dataset[_],
+      featuresCol: String,
+      labelCol: String): Array[SelectionTestResult] = {
+
+    val spark = dataset.sparkSession
+    import spark.implicits._
+
+    SchemaUtils.checkColumnType(dataset.schema, featuresCol, new VectorUDT)
+    SchemaUtils.checkNumericType(dataset.schema, labelCol)
+
+    val labeledPointRdd = dataset.select(col(labelCol).cast("double"), col(featuresCol))
+      .as[(Double, Vector)]
+      .rdd.map { case (label, features) => LabeledPoint(label, features) }
+
+    val numFeatures = MetadataUtils.getNumFeatures(dataset, featuresCol)
+    val Row(numSamples: Long, numClasses: Long) =
+      dataset.select(count(labelCol), countDistinct(labelCol)).head
+
+    labeledPointRdd.flatMap { case LabeledPoint(label, features) =>
+      features.iterator.map { case (col, value) =>
+        (col, (label, value, value * value))
+      }
+    }.aggregateByKey[(Double, Double, OpenHashMap[Double, Double], OpenHashMap[Double, Long])](
+      (0.0, 0.0, new OpenHashMap[Double, Double], new OpenHashMap[Double, Long]))(
+      seqOp = {
+        case (
+          (sum: Double, sumOfSq: Double, mapOfSumPerClass, mapOfCountPerClass),
+          (label, feature, featureSq)
+         ) =>
+          mapOfSumPerClass.changeValue(label, feature, _ + feature)
+          mapOfCountPerClass.changeValue(label, 1L, _ + 1L)
+          (mapOfSumPerClass, mapOfCountPerClass)
+          (sum + feature, sumOfSq + featureSq, mapOfSumPerClass, mapOfCountPerClass)
+      },
+      combOp = {
+        case (
+          (sum1, sumOfSq1, mapOfSumPerClass1, mapOfCountPerClass1),
```

Comment:
nit, I guess we can use short variable names to make this more concise， like
```scala
combOp = { case (sum1, sqSum1, sums1, counts1), (sum2, sqSum2, sums2, counts2) =>
  ...
}

```

Reply:
I changed ```mapOfSumPerClass``` and ```mapOfCountPerClass```, but didn't change ```sumOfSq```. I have 
```val sqSum = sum * sum```